### PR TITLE
dzx-dzx/issue28

### PR DESCRIPTION
**Pull request by dzx-dzx**

### CODE REVIEW

**Comment by callstackexceed** [comment]
There are bugs at line 53 and 71, I recommed to fix it as soon as possible.

**Comment by callstackexceed** [comment]
By the way, line 53 and 71 may throw Expection unexpectly.

**Comment by callstackexceed** [comment]
My recommand
```patch
diff --git a/constructor.js b/constructor.js
index ed845e7..a72fe61 100644
--- a/constructor.js
+++ b/constructor.js
@@ -50,7 +50,7 @@ class BlockType {
             return this._data.blockstate;
         }
         else {
-            if (!_blockMap) return null
+            if (!BlockType._blockMap) return null
             const { id, data } = BlockType._blockMap.get(this.blockIdentifier)
             if (!id || !data.hasOwnProperty(this.tiledata)) return null
             this._data.blockstate = data[this.tiledata]
@@ -68,7 +68,7 @@ class BlockType {
             return this._data.tiledata
         }
         else {
-            if (!_blockMap) return null
+            if (!BlockType._blockMap) return null
             const { id, data } = BlockType._blockMap.get(this.blockIdentifier)
             if (!id) return null
             for (const tiledata in data)
```

**Comment by callstackexceed** [comment]
My recommand on removing class field:
```patch
diff --git a/constructor.js b/constructor.js
index ed845e7..af21726 100644
--- a/constructor.js
+++ b/constructor.js
@@ -25,11 +25,11 @@ function assertBlockstateEqual(a, b) {
     return a === b || (a && b && typeof a === "object" && typeof b === "object" && Object.keys(a).length === Object.keys(b) && Object.keys(a).every((property) => a[property] === b[property]))
 }
 class BlockType {
-    _data = { "blockIdentifier": null, "blockstate": null, "tiledata": null }
-    _flag = { "blockstateUpToDate": false, "tiledataUpToDate": false }
     static _blockMap = null
     constructor(blockIdentifier) {
         this.blockIdentifier = blockIdentifier;
+        this._data = { "blockIdentifier": null, "blockstate": null, "tiledata": null };
+        this._flag = { "blockstateUpToDate": false, "tiledataUpToDate": false };
     }
     static fromBlockstate(blockIdentifier, blockstate) {
         let blockType = new BlockType(blockIdentifier);
```

File: constructor.js [hunk 1]
```diff
@@ -9,17 +9,90 @@ class Coordinate {
...
+function assertBlockstateEqual(a, b) {
+    return a === b || (a && b && typeof a === "object" && typeof b === "object" && Object.keys(a).length === Object.keys(b) && Object.keys(a).every((property) => a[property] === b[property]))
+}
 class BlockType {
-    constructor(blockIdentifier, blockState) {
+    _data = { "blockIdentifier": null, "blockstate": null, "tiledata": null }
+    _flag = { "blockstateUpToDate": false, "tiledataUpToDate": false }
+    static _blockMap = null
+    constructor(blockIdentifier) {
         this.blockIdentifier = blockIdentifier;
```
Comment (dzx-dzx) [inline constructor.js:32]: As per @callstackexceed 's suggestion.

**Comment by callstackexceed** [comment]
> 
> 
> By the way, line 53 and 71 may throw Expection unexpectly.

My recommand on it:

```patch
diff --git a/constructor.js b/constructor.js
index 85f1313..0c8b3f2 100644
--- a/constructor.js
+++ b/constructor.js
@@ -51,7 +51,7 @@ class BlockType {
         }
         else {
             if (!BlockType._blockMap) return null
-            const { id, data } = BlockType._blockMap.get(this.blockIdentifier)
+            const { id, data } = BlockType._blockMap.get(this.blockIdentifier) || {};
             if (!id || !data.hasOwnProperty(this.tiledata)) return null
             this._data.blockstate = data[this.tiledata]
             this._flag.blockstateUpToDate = true
@@ -69,7 +69,7 @@ class BlockType {
         }
         else {
             if (!BlockType._blockMap) return null
-            const { id, data } = BlockType._blockMap.get(this.blockIdentifier)
+            const { id, data } = BlockType._blockMap.get(this.blockIdentifier) || {};
             if (!id) return null
             for (const tiledata in data)
                 if (assertBlockstateEqual(data[tiledata], this.blockstate)) {
```

**Comment by callstackexceed** [comment]
Remove `??` in script.
```patch
diff --git a/constructor.js b/constructor.js
index 85f1313..4cfe8eb 100644
--- a/constructor.js
+++ b/constructor.js
@@ -16,9 +16,9 @@ class Coordinate {
     }
 }
 class Position {
-    constructor(coordinate, dimension) {
+    constructor(coordinate, dimension = "overworld") {
         this.coordinate = coordinate;
-        this.dimension = dimension ?? "overworld";
+        this.dimension = dimension;
     }
 }
 function assertBlockstateEqual(a, b) {
```

**Comment by callstackexceed** [comment]
```patch
diff --git a/constructor.js b/constructor.js
index 88ec733..a8cf10f 100644
--- a/constructor.js
+++ b/constructor.js
@@ -26,10 +26,21 @@ function assertBlockstateEqual(a, b) {
 }
 class BlockType {
     static _blockMap = null
-    constructor(blockIdentifier) {
+    constructor(blockIdentifier, blockstateOrTiledata) {
         this.blockIdentifier = blockIdentifier;
         this._data = { "blockIdentifier": null, "blockstate": null, "tiledata": null };
         this._flag = { "blockstateUpToDate": false, "tiledataUpToDate": false };
+        switch(typeof blockstateOrTiledata) {
+        case "object":
+            this.blockstate = blockstateOrTiledata;
+            break;
+        case "number":
+            this.tiledata = blockstateOrTiledata;
+            break;
+        case "undefined":
+        default:
+            break;
+        }
     }
     static fromBlockstate(blockIdentifier, blockstate) {
         let blockType = new BlockType(blockIdentifier);
```

**Comment by callstackexceed** [comment]
```patch
diff --git a/constructor.js b/constructor.js
index 88ec733..3887a32 100644
--- a/constructor.js
+++ b/constructor.js
@@ -49,13 +49,15 @@ class BlockType {
         if (this._flag.blockstateUpToDate) {
             return this._data.blockstate;
         }
-        else {
+        else if (this._flag.tiledataUpToDate) {
             if (!BlockType._blockMap) return null
             const { id, data } = BlockType._blockMap.get(this.blockIdentifier) || {}
             if (!id || !data.hasOwnProperty(this.tiledata)) return null
             this._data.blockstate = data[this.tiledata]
             this._flag.blockstateUpToDate = true
             return this.blockstate
+        } else {
+            return {};
         };
     }
     set blockstate(blockstate) {
@@ -67,7 +69,7 @@ class BlockType {
         if (this._flag.tiledataUpToDate) {
             return this._data.tiledata
         }
-        else {
+        else if (this._flag.blockstateUpToDate) {
             if (!BlockType._blockMap) return null
             const { id, data } = BlockType._blockMap.get(this.blockIdentifier) || {}
             if (!id) return null
@@ -77,6 +79,8 @@ class BlockType {
                     return tiledata
                 }
             return null;
+        } else {
+            return 0;
         };
     }
     set tiledata(tiledata) {
```

**Comment by callstackexceed** [comment]
```patch
diff --git a/constructor.js b/constructor.js
index 88ec733..d91075a 100644
--- a/constructor.js
+++ b/constructor.js
@@ -26,10 +26,21 @@ function assertBlockstateEqual(a, b) {
 }
 class BlockType {
     static _blockMap = null
-    constructor(blockIdentifier) {
+    constructor(blockIdentifier, blockstateOrTiledata) {
         this.blockIdentifier = blockIdentifier;
         this._data = { "blockIdentifier": null, "blockstate": null, "tiledata": null };
         this._flag = { "blockstateUpToDate": false, "tiledataUpToDate": false };
+        switch(typeof blockstateOrTiledata) {
+        case "object":
+            this.blockstate = blockstateOrTiledata;
+            break;
+        case "number":
+            this.tiledata = blockstateOrTiledata;
+            break;
+        case "undefined":
+            this.tiledata = 0;
+        default:
+            break;
     }
     static fromBlockstate(blockIdentifier, blockstate) {
         let blockType = new BlockType(blockIdentifier);
```

**Comment by dzx-dzx** [comment]
A research into the mapping of blockstate and tiledata suggests that for certain blocks the blockType would need its blockIdentidier changed if converted to tiledata(example: `minecraft:wood`). And actually, we doubt the significance of the conversion feature.
Thus I recommend to drop all changes concerning BlockType, and this pr is freezed before further discussion.

**Comment by dzx-dzx** [comment]
Oh, and, the new `Position` has its property `tickingArea` removed. It shall be saved in user's `session`.

**Comment by callstackexceed** [comment]
> 
> 
> Oh, and, the new `Position` has its property `tickingArea` removed. It shall be saved in user's `session`.

This will be a breaking change, I think.

**Comment by callstackexceed** [comment]
> 
> 
> A research into the mapping of blockstate and tiledata suggests that for certain blocks the blockType would need its blockIdentidier changed if converted to tiledata(example: `minecraft:wood`). And actually, we doubt the significance of the conversion feature.
> Thus I recommend to drop all changes concerning BlockType, and this pr is freezed before further discussion.

OK. You can temporarily close this if you want.

**Comment by dzx-dzx** [comment]
As additional block data such as nbt can be obtained in some backends like lxl, we shall take it into consideration that an additional space would be needed to store arbitrary data.

**Comment by dzx-dzx** [comment]
We have decided on Blocktype only nbt will be stored, if could.

**Comment by dzx-dzx** [comment]
@callstackexceed

**Review by callstackexceed — APPROVED** [review]
LGTM.

**Pull request merged**